src/MongoDB/TopologyDescription.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <Zend/zend_interfaces.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+
+zend_class_entry* php_phongo_topologydescription_ce;
+
+#define PHONGO_TOPOLOGY_UNKNOWN "Unknown"
+#define PHONGO_TOPOLOGY_SINGLE "Single"
+#define PHONGO_TOPOLOGY_SHARDED "Sharded"
+#define PHONGO_TOPOLOGY_REPLICA_SET_NO_PRIMARY "ReplicaSetNoPrimary"
+#define PHONGO_TOPOLOGY_REPLICA_SET_WITH_PRIMARY "ReplicaSetWithPrimary"
+
+/* {{{ proto array MongoDB\Driver\TopologyDescription::getServers()
+   Returns an array of ServerDescription objects for all known servers in the topology */
+static PHP_METHOD(TopologyDescription, getServers)
+{
+	php_phongo_topologydescription_t* intern;
+	mongoc_server_description_t**     sds;
+	size_t                            i, n = 0;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	sds = mongoc_topology_description_get_servers(intern->topology_description, &n);
```

Comment:
This appears to leak, but not because of the server descriptions themselves. Those are used to initialize PHP objects and ultimately get freed in ServerDescription's `free_object` handler. Rather, the allocated array that stores the server description pointers is leaking here. If you look at `mongoc_server_descriptions_destroy_all` in libmongoc, you'll see that it frees both the pointers and finally the array.

Two possible solutions here:

 * Use `mongoc_server_description_new_copy` in the `for` loop below to send copies to `phongo_serverdescription_init`, which will be owned by the PHP object. And then call `mongoc_server_descriptions_destroy_all` after the loop.

 * Avoid calling `mongoc_server_descriptions_destroy_all` and making extra copies of the server descriptions and just `bson_free(sds)` after the `for` loop.

I'd lean towards the first even though it's a bit more inefficient, since knowing to free to array alone is kind of assuming libmongoc's internal logic and not abiding its public API.